### ISSUE TITLE
Push error onto the stack when malloc fails in scrypt

### ISSUE THREAD

**Issue by alex** [openssl/openssl#5781]
https://github.com/openssl/openssl/blob/master/crypto/evp/pbe_scrypt.c#L232-L234

It's currently very difficult to debug if you get a malloc failure from scrypt because nothing is pushed onto the openssl error stack

**Comment by mspncp** [comment]
Whoever fixes this issue might want to skim over the other locations where no error is pushed after a malloc failure. Maybe one or the other location deserves an error message, too. To find them, just run the following command from the OpenSSL source directory:
```
pcregrep -rnM 'OPENSSL_.?alloc.*\n.*if.*NULL.*\n.*return'  crypto ssl
```

**Comment by richsalz** [comment]
It would be nice if we could push the error down into the CRYPTO_alloc routines. We'd lose the function name, but we could preserve the source file/line numbers.

**Comment by mspncp** [comment]
You mean, preserve file/line numbers of the caller, right? That sounds like a good idea. Initially, I also thought about pushing the error in CRYPTO_*alloc() itself, but then I discarded that thought because the line information of the alloc function itself would have been practically useless.

**Comment by richsalz** [comment]
Look at the define's for OPENSSL_malloc and friends, and the body of CRYPTO_malloc and friends.
Would still lose the function name tho.

**Comment by mspncp** [comment]
I‘ll think about it. But for today I‘m offline now.

**Comment by bernd-edlinger** [comment]
Pushing something on the error stack is likely to fail when there is no more memory left.

**Comment by bernd-edlinger** [comment]
OTOH this function seems to have numerous error exits without instrumentation.

**Comment by alex** [comment]
Scrypt can use substaintal amounts of memory, I think it's likely that it
may cases an error stack allocation can succeed when the scrypt allocation
failed

**Comment by FdaSilvaYY** [comment]
@bernd-edlinger : there is no allocation in ERR_put_error function. storage is allocated at startup.
@mspncp : your grep is retrieving 87 occurrences ...

**Comment by mspncp** [comment]
I know. That‘s why I didn‘t post them ;-)

**Comment by bernd-edlinger** [comment]
Well @FdaSilvaYY maybe not in ERR_put_error, but look at ERR_get_state.

**Comment by FdaSilvaYY** [comment]
@bernd-edlinger : you are right, it better that the app push an error as soon as possible to init the error-storage start ERR_STATE, or call directly ERR_get_state from each thread startup.
I remenber that  CRL loading trigger systematically some errors ...  
May-be add a word about this in documentation ?
---
BTW this issue is fixed by 3484236d8d7afedd3e5c7771bd49d3385340e3bf

**Comment by mspncp** [comment]
> BTW this issue is fixed in master : 3484236

This issue was not closed automatically, because the commit did not contain a 'Fixes #5781' line. Never mind, I'll close it now.

**Comment by DDvO** [comment]
Further, more general discussion on this topic has been started in issue #6251.
